pypsa/optimization/constraints.py

```diff
@@ -105,16 +154,391 @@ def define_operational_constraints_for_extendables(
+
+    If the number of modules is not an integer number, the function returns an error message.
+    _____
+
+    Parameters
+    ----------
+    n : pypsa.Network
+    sns : pd.Index
+        Snapshots of the constraint.
+    c : str
+        name of the network component
+    attr : str
+        name of the attribute, e.g. 'p_nom'
+    """
+    com_i = n.get_committable_i(c)
+    fix_i = n.df(c).query(f"not {attr}_extendable").index
+    mod_i = n.df(c).query(f"({attr}_mod>0)").index
+
+    inter_i = com_i.intersection(mod_i).intersection(fix_i).rename(com_i.name)
+
+    if inter_i.empty:
+        return
+
+    m = n.model
+
+    ext_i = n.get_extendable_i(c)
+    active = get_activity_mask(n, c, sns, ext_i) if n._multi_invest else None
+
+    n_mod = n.df(c)[attr].loc[inter_i] / n.df(c)[attr + "_mod"].loc[inter_i]
+    diff_n_mod = abs(n_mod - round(n_mod))
+    non_integers_n_mod_i = diff_n_mod[diff_n_mod > 10**-6].index
+
+    if not non_integers_n_mod_i.empty:
+        msg = (
+            "For non-extendable but committable assets, if both p_nom and p_nom_mod are declared, p_nom must be a"
+            "must be a multiple of p_nom_mod. Found "
+            f"assets in component {c} do not respect this criterion:"
+            f"\n\n\t{', '.join(non_integers_n_mod_i)}"
+        )
+        raise ValueError(msg)
+
+    rhs = pd.DataFrame(0, sns, inter_i)
+    rhs.loc[sns, inter_i] = n_mod.loc[inter_i].values
+
+    status = m.variables[f"{c}-status"].loc[sns, inter_i]
+    n.model.add_constraints(
+        status, "<=", rhs, name=f"{c}-status-{attr}-upper-mod_nonext", mask=active
+    )
+
+    start_up = m.variables[f"{c}-start_up"].loc[sns, inter_i]
+    n.model.add_constraints(
+        start_up, "<=", rhs, name=f"{c}-start_up-{attr}-upper-mod_nonext", mask=active
+    )
+
+    shut_down = m.variables[f"{c}-shut_down"].loc[sns, inter_i]
+    n.model.add_constraints(
+        shut_down, "<=", rhs, name=f"{c}-shut_down-{attr}-upper-mod_nonext", mask=active
+    )
+
+
+def define_committability_variables_constraints_for_modular_and_extendables(
+    n, sns, c, attr
+):
+    """
+    This function sets the upper limit of committable variables (status, start-up, shut-down) to the
+    variable n_mod for all committable, extendable and modular components.
+    _____
+
+    Parameters
+    ----------
+    n : pypsa.Network
+    sns : pd.Index
+        Snapshots of the constraint.
+    c : str
+        name of the network component
+    attr : str
+        name of the attribute, e.g. 'p_nom'
+    """
+    com_i = n.get_committable_i(c)
+    ext_i = n.df(c).query(f"{attr}_extendable").index
+    mod_i = n.df(c).query(f"({attr}_mod>0)").index
+
+    inter_i = com_i.intersection(mod_i).intersection(ext_i).rename(com_i.name)
+
+    if inter_i.empty:
+        return
+
+    m = n.model
+
+    active = get_activity_mask(n, c, sns, ext_i) if n._multi_invest else None
+
+    n_mod = n.model[f"{c}-n_mod"].loc[inter_i]
+
+    # ALLERT DIMENSIONI DISEQUAZIONE
```

Comment:
Can you explain these comments?

Reply:
I think they comments for my self that I forgot to remove xD